File: libs/ts/src/credentials/format.ts
/**
 * Vault on-disk format: model types + the normative byte constructions (AEAD AADs and the
 * length-prefixed canonical MAC input), matching the Rust/Python/Java references. The MAC is over
 * this byte string, not the JSON text, so JSON formatting may differ across languages.
 */
export const FORMAT_VERSION = 1;

export interface KekInfo {
  provider: string;
  alg: string;
  wrapNonce?: string;
  wrappedDek: string;
  kmsKeyId?: string;
}

export interface VersionEntry {
  version: string;
  createdMs: number;
  ttlSecs?: number;
  source: string;
  centralVersionId?: string;
  labels?: Record<string, string>;
  contentType?: string;
  nonce: string;
  ciphertext: string;
}

export interface SecretEntry {
  versions: VersionEntry[];
}

export interface VaultFile {
  format: number;
  vaultId: string;
  kek: KekInfo;
  secrets: Record<string, SecretEntry>;
  mac: string;
}

/** AEAD AAD binding a record to its vault, name, and version. */
export function recordAad(vaultId: string, name: string, version: string): Buffer {
  return Buffer.from(`ggcommons-vault/v1|${vaultId}|${name}|${version}`, "utf8");
}

/** AEAD AAD binding the wrapped DEK to its vault. */
export function dekWrapAad(vaultId: string): Buffer {
  return Buffer.from(`ggcommons-vault/v1/dek-wrap|${vaultId}`, "utf8");
}

function lp(b: Buffer): Buffer {
  const len = Buffer.alloc(4);
  len.writeUInt32LE(b.length);
  return Buffer.concat([len, b]);
}

function u32le(n: number): Buffer {
  const b = Buffer.alloc(4);
  b.writeUInt32LE(n);
  return b;
}

function u64le(n: number): Buffer {
  const b = Buffer.alloc(8);
  b.writeBigUInt64LE(BigInt(n));
  return b;
}

/**
 * Build the canonical MAC input over the whole secret set. Secrets are ordered by their UTF-8 name
 * bytes (unsigned, via `Buffer.compare`) to match Rust's `BTreeMap` / Python's sort. Layout: see
 * the Rust reference `mac_input`.
 */
export function macInput(
  vaultId: string,
  secrets: Record<string, SecretEntry>,
  decodeB64: (s: string) => Buffer,
): Buffer {
  const parts: Buffer[] = [
    Buffer.from("ggcommons-vault/v1/mac", "utf8"),
    lp(Buffer.from(vaultId, "utf8")),
  ];
  const names = Object.keys(secrets).sort((a, b) =>
    Buffer.compare(Buffer.from(a, "utf8"), Buffer.from(b, "utf8")),
  );
  parts.push(u32le(names.length));
  for (const name of names) {
    parts.push(lp(Buffer.from(name, "utf8")));
    const versions = secrets[name].versions;
    parts.push(u32le(versions.length));
    for (const v of versions) {
      parts.push(lp(Buffer.from(v.version, "utf8")));
      parts.push(u64le(v.createdMs ?? 0));
      parts.push(u64le(v.ttlSecs ?? 0));
      parts.push(lp(Buffer.from(v.source ?? "", "utf8")));
      parts.push(lp(Buffer.from(v.centralVersionId ?? "", "utf8")));
      parts.push(lp(decodeB64(v.nonce)));
      parts.push(lp(decodeB64(v.ciphertext)));
    }
  }
  return Buffer.concat(parts);
}

File: libs/rust/src/credentials/format.rs
//! # Vault on-disk format (normative — identical across all language ports)
//!
//! **One-liner purpose**: The serde model of the vault file plus the exact byte constructions
//! (AEAD AAD and the MAC input) that every binding must reproduce.
//!
//! ## Overview
//! The vault is a single JSON file (see [`VaultFile`]). Numbers that are encrypted are stored as
//! base64; the secret name map is a [`BTreeMap`] so JSON key order is deterministic. Integrity is
//! a single HMAC ([`VaultFile::mac`]) over a **length-prefixed canonical byte string** (not the
//! JSON text) so it is insensitive to JSON formatting differences between languages.
//!
//! ## Semantics & Architecture
//! - `format = 1`. `vaultId` is a stable per-vault UUID, bound into every AAD and the MAC so
//!   records cannot be copied between vaults.
//! - **Record AAD** (`record_aad`): `ggcommons-vault/v1|<vaultId>|<name>|<version>` — binds each
//!   ciphertext to its identity *and* version (anti-swap, anti in-file rollback of a record).
//! - **DEK-wrap AAD** (`dek_wrap_aad`): `ggcommons-vault/v1/dek-wrap|<vaultId>`.
//! - **MAC input** (`mac_input`): length-prefixed concatenation of the full secret set (raw
//!   nonce/ciphertext bytes, not base64); see the function for the exact layout. The MAC key is
//!   `HKDF-SHA256(DEK, salt=vaultId, info="…/mac")`.
//!
//! ## Design choices
//! Length-prefixed binary MAC input (rather than canonical JSON) is how interoperable encrypted
//! formats (age, JWE) stay byte-identical across independent implementations without sharing code.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

/// Current on-disk format version.
pub const FORMAT_VERSION: u32 = 1;

/// The whole vault file.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VaultFile {
    pub format: u32,
    pub vault_id: String,
    pub kek: KekInfo,
    /// Secret name → entry. `BTreeMap` keeps JSON keys sorted (deterministic output).
    pub secrets: BTreeMap<String, SecretEntry>,
    /// base64(HMAC-SHA256) over [`mac_input`].
    pub mac: String,
}

/// How the DEK is wrapped — written by the [`super::keyprovider::KeyProvider`].
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct KekInfo {
    /// Custodian id: `file` | `kms` | `greengrass` | `pkcs11` | `env`.
    pub provider: String,
    /// Wrap algorithm, e.g. `AES-256-GCM` (file/env) or `aws-kms` (kms).
    pub alg: String,
    /// base64 nonce used to wrap the DEK (AEAD wrap providers).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub wrap_nonce: Option<String>,
    /// base64 wrapped DEK (AEAD ciphertext, or the KMS ciphertext blob).
    pub wrapped_dek: String,
    /// KMS key id/arn (provider = kms/greengrass).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub kms_key_id: Option<String>,
}

/// All retained versions of one secret (newest last).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SecretEntry {
    pub versions: Vec<VersionEntry>,
}

/// One encrypted version of a secret.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VersionEntry {
    /// Monotonic, zero-padded 8-digit version (e.g. `00000003`).
    pub version: String,
    pub created_ms: u64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ttl_secs: Option<u64>,
    /// `local` (written via `put`) or `central` (pulled from the upstream vault).
    pub source: String,
    /// Upstream version id for change detection (central source).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub central_version_id: Option<String>,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub labels: BTreeMap<String, String>,
    #[serde(default = "default_content_type")]
    pub content_type: String,
    /// base64 AEAD nonce for this version.
    pub nonce: String,
    /// base64 AEAD `ciphertext || tag`.
    pub ciphertext: String,
}

fn default_content_type() -> String {
    "application/octet-stream".to_string()
}

/// AAD binding a record's ciphertext to its vault, name, and version.
pub fn record_aad(vault_id: &str, name: &str, version: &str) -> Vec<u8> {
    format!("ggcommons-vault/v1|{vault_id}|{name}|{version}").into_bytes()
}

/// AAD binding the wrapped DEK to its vault.
pub fn dek_wrap_aad(vault_id: &str) -> Vec<u8> {
    format!("ggcommons-vault/v1/dek-wrap|{vault_id}").into_bytes()
}

/// Build the canonical MAC input over the whole secret set.
///
/// # Semantics
/// Layout (all integers little-endian; `lp(x)` = `u32_le(len) ‖ x`):
/// ```text
/// b"ggcommons-vault/v1/mac"
///   ‖ lp(vaultId)
///   ‖ u32_le(secret_count)
///   ‖ for each secret (BTreeMap order = name byte order):
///       lp(name) ‖ u32_le(version_count)
///         ‖ for each version (array order):
///             lp(version) ‖ u64_le(createdMs) ‖ u64_le(ttlSecs or 0)
///             ‖ lp(source) ‖ lp(centralVersionId or "")
///             ‖ lp(nonce_raw) ‖ lp(ciphertext_raw)
/// ```
/// `nonce_raw`/`ciphertext_raw` are the **decoded** bytes (the MAC is over raw bytes, not the
/// base64 text). Authenticates every field that matters; insensitive to JSON formatting.
pub fn mac_input(
    vault_id: &str,
    secrets: &BTreeMap<String, SecretEntry>,
    decode_b64: impl Fn(&str) -> Vec<u8>,
) -> Vec<u8> {
    fn lp(out: &mut Vec<u8>, bytes: &[u8]) {
        out.extend_from_slice(&(bytes.len() as u32).to_le_bytes());
        out.extend_from_slice(bytes);
    }
    let mut out = Vec::with_capacity(256);
    out.extend_from_slice(b"ggcommons-vault/v1/mac");
    lp(&mut out, vault_id.as_bytes());
    out.extend_from_slice(&(secrets.len() as u32).to_le_bytes());
    for (name, entry) in secrets {
        lp(&mut out, name.as_bytes());
        out.extend_from_slice(&(entry.versions.len() as u32).to_le_bytes());
        for v in &entry.versions {
            lp(&mut out, v.version.as_bytes());
            out.extend_from_slice(&v.created_ms.to_le_bytes());
            out.extend_from_slice(&v.ttl_secs.unwrap_or(0).to_le_bytes());
            lp(&mut out, v.source.as_bytes());
            lp(&mut out, v.central_version_id.as_deref().unwrap_or("").as_bytes());
            lp(&mut out, &decode_b64(&v.nonce));
            lp(&mut out, &decode_b64(&v.ciphertext));
        }
    }
    out
}
